homeassistant/components/downloader/__init__.py

```diff
@@ -57,122 +87,137 @@ def setup(hass: HomeAssistant, config: ConfigType) -> bool:
+            path, ext = os.path.splitext(final_path)
 
-                    with open(final_path, "wb") as fil:
-                        for chunk in req.iter_content(1024):
-                            fil.write(chunk)
+            # If file exist append a number.
+            # We test filename, filename_2..
+            if not overwrite:
+                tries = 1
+                final_path = path + ext
+                while os.path.isfile(final_path):
+                    tries += 1
 
-                    _LOGGER.debug("Downloading of %s done", url)
-                    hass.bus.fire(
-                        f"{DOMAIN}_{DOWNLOAD_COMPLETED_EVENT}",
-                        {"url": url, "filename": filename},
-                    )
+                    final_path = f"{path}_{tries}.{ext}"
 
-            except requests.exceptions.ConnectionError:
-                _LOGGER.exception("ConnectionError occurred for %s", url)
-                hass.bus.fire(
-                    f"{DOMAIN}_{DOWNLOAD_FAILED_EVENT}",
-                    {"url": url, "filename": filename},
-                )
+            _LOGGER.debug("%s -> %s", url, final_path)
 
-                # Remove file if we started downloading but failed
-                if final_path and os.path.isfile(final_path):
-                    os.remove(final_path)
-            except ValueError:
-                _LOGGER.exception("Invalid value")
-                hass.bus.fire(
-                    f"{DOMAIN}_{DOWNLOAD_FAILED_EVENT}",
-                    {"url": url, "filename": filename},
-                )
+            with open(final_path, "wb") as fil:
+                for chunk in req.iter_content(1024):
+                    fil.write(chunk)
 
-                # Remove file if we started downloading but failed
-                if final_path and os.path.isfile(final_path):
-                    os.remove(final_path)
+            _LOGGER.debug("Downloading of %s done", url)
+            hass.bus.fire(
+                f"{DOMAIN}_{DOWNLOAD_COMPLETED_EVENT}",
+                {"url": url, "filename": filename},
+            )
 
-        threading.Thread(target=do_download).start()
+    except requests.exceptions.ConnectionError:
+        _LOGGER.exception("ConnectionError occurred for %s", url)
+        hass.bus.fire(
+            f"{DOMAIN}_{DOWNLOAD_FAILED_EVENT}",
+            {"url": url, "filename": filename},
+        )
 
-    hass.services.register(
-        DOMAIN,
-        SERVICE_DOWNLOAD_FILE,
-        download_file,
-        schema=SERVICE_DOWNLOAD_FILE_SCHEMA,
-    )
+        # Remove file if we started downloading but failed
+        if final_path and os.path.isfile(final_path):
+            os.remove(final_path)
+    except ValueError:
+        _LOGGER.exception("Invalid value")
+        hass.bus.fire(
+            f"{DOMAIN}_{DOWNLOAD_FAILED_EVENT}",
+            {"url": url, "filename": filename},
+        )
 
-    return True
```

Comment:
Not entirely sure if you mean this method, or something else? Am I overlooking a part?

```
def download_file(
    hass: HomeAssistant, service: ServiceCall, download_path: str
) -> None:
    """Start thread to download file specified in the URL."""
    threading.Thread(target=do_download, args=(hass, service, download_path)).start()```

Reply:
I mean that `download_file` was previously an inner function of `setup`, but you moved it to class level. Since that has nothing to do with if the `downloader` integration is setup via config flow I'm asking to revert that change such that `download_file` is instead an inner function of `async_setup_entry`.